LOOP-488 Bolus progress and quick cancellation

ps2 opened the pull request.

ps2 committed:
Bolus progress and quick cancellation

ps2 committed:
Add MKRingProgressView files directly

ps2 committed:
Update LICENCE file to include license for MKRingProgressView

ps2 committed:
Use MKRingProgressView via carthage

ps2 committed:
Use pumpmanager's bolusProgressEstimator

mpangburn commented on Loop/View Controllers/StatusTableViewController.swift:
```diff
@@ -16,24 +16,19 @@ import SwiftCharts
 
     private let log = OSLog(category: "StatusTableViewController")
 
+    lazy var quantityFormatter: QuantityFormatter = {
```
`lazy var quantityFormatter = QuantityFormatter()`

mpangburn commented on Loop/View Controllers/StatusTableViewController.swift:
```diff
@@ -713,17 +735,24 @@ final class StatusTableViewController: ChartsTableViewController {
-            let cell = tableView.dequeueReusableCell(withIdentifier: TitleSubtitleTableViewCell.className, for: indexPath) as! TitleSubtitleTableViewCell
-            cell.selectionStyle = .none
+
+            let getTitleSubtitleCell: () -> TitleSubtitleTableViewCell = {
```
Clearer as a `func`

mpangburn commented on Loop/Views/BolusProgressTableViewCell.swift:
```diff
@@ -0,0 +1,103 @@
+
+    @IBOutlet weak var stopSquare: UIView! {
+        didSet {
+            stopSquare.layer.cornerRadius = 3
```
I think this corner radius is _slightly_ too large. I played briefly with changing it and found that a value of 2 preserves the square shape more clearly.

mpangburn commented on Loop/Views/BolusProgressTableViewCell.swift:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private func updateProgress() {
+        if let totalUnits = totalUnits, let unit = unit {
```
Nit: `guard` here

mpangburn commented on Loop/Views/TitleSubtitleTableViewCell.swift:
```diff
@@ -12,6 +12,8 @@ class TitleSubtitleTableViewCell: UITableViewCell {
 
     @IBOutlet weak var titleLabel: UILabel!
 
+    @IBOutlet weak var progressIndicator: UIProgressView!
```
Is this a remnant from a previous UI implementation?

mpangburn reviewed (approved):
Looks really nice. Just a couple UI fixes:
- The status row hides before the ring closing animation finishes for a completed bolus.
- When another bolus starts after one has completed, the full ring's value is visibly unwound when the status row appears. The ring should be reset before the user sees it again.

darinkrauss commented on Loop/Views/BolusProgressTableViewCell.swift:
```diff
@@ -0,0 +1,103 @@
+
+    override public func prepareForReuse() {
+        super.prepareForReuse()
+        progressIndicator.progress = 0
```
`progressLabel.text = ""`?

darinkrauss reviewed (approved):
LGTM 🚀

ps2 committed:
Updates from code review:

- Delay removal of bolus progress so user sees progress finish
- Reset ring progress during cell re-use
- Tweak cornerRadius
- Code clarity updates

ps2 committed:
PumpManager Protocol updates

ps2 committed:
Protocol naming update

ps2 committed:
Create bolus progress reporter

mpangburn reviewed (approved):
Nothing to add since the last review. 🚀

ps2 committed:
Fix progress animating from 1 to 0 on reuse.

darinkrauss reviewed (approved):
LGTM (again) 🏎
